event/oracle/upgrade_oracle_event.go

```diff
@@ -0,0 +1,136 @@
+	}
+
+	log.Infof("oracle upgrade approval info. unique ID(%s), approver address(%s), target address(%s)",
+		msgApproveOracleUpgrade.ApprovalSharingOracleKey.UniqueId,
+		msgApproveOracleUpgrade.ApprovalSharingOracleKey.ApproverOracleAddress,
+		msgApproveOracleUpgrade.ApprovalSharingOracleKey.TargetOracleAddress,
+	)
+
+	txHeight, txHash, err := e.reactor.BroadcastTx(msgApproveOracleUpgrade)
+	if err != nil {
+		return fmt.Errorf("failed to ApproveOracleUpgrade transaction for oracle upgrade: %w", err)
+	}
+
+	log.Infof("succeeded to ApproveOracleUpgrae transaction for oracle upgrade. height(%d), hash(%s)", txHeight, txHash)
+
+	return nil
+}
+
+func (e UpgradeOracleEvent) verifyOracleUpgrade(ctx context.Context, oracleUpgrade *oracletypes.OracleUpgrade, uniqueID, targetAddress string) error {
+	queryClient := e.reactor.QueryClient()
+
+	oracleUpgradeInfo, err := queryClient.GetOracleUpgradeInfo(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get oracle upgrade info: %w", err)
+	}
+
+	// check if the unique ID is same with the one stored in panacea
+	if uniqueID != oracleUpgradeInfo.GetUniqueId() {
+		return fmt.Errorf("the upgrade unique ID is different from the one stored in panacea")
+	}
+
+	// check if the oracle is registered
+	if _, err := queryClient.GetOracle(ctx, targetAddress); err != nil {
+		return fmt.Errorf("failed to get oracle: %w", err)
+	}
+
+	// verify trusted block info
+	if err := queryClient.VerifyTrustedBlockInfo(oracleUpgrade.TrustedBlockHeight, oracleUpgrade.TrustedBlockHash); err != nil {
+		return fmt.Errorf("failed to verify trusted block information. height(%d), hash(%s): %w", oracleUpgrade.TrustedBlockHeight, oracleUpgrade.TrustedBlockHash, err)
+	}
+
+	// verify remote report
+	nodePubKeyHash := sha256.Sum256(oracleUpgrade.NodePubKey)
+
+	uniqueIDBz, err := hex.DecodeString(oracleUpgradeInfo.GetUniqueId())
+	if err != nil {
+		return fmt.Errorf("failed to decode unique ID: %w", err)
+	}
+
+	if err := sgx.VerifyRemoteReport(oracleUpgrade.GetNodePubKeyRemoteReport(), nodePubKeyHash[:], uniqueIDBz); err != nil {
+		return fmt.Errorf("failed to verify remote report: %w", err)
+	}
+
+	return nil
+}
+
+func (e UpgradeOracleEvent) generateApproveOracleUpgradeMsg(oracleUpgrade *oracletypes.OracleUpgrade, targetAddress string) (*oracletypes.MsgApproveOracleUpgrade, error) {
+	approverAddress := e.reactor.OracleAcc().GetAddress()
+	oraclePrivKeyBz := e.reactor.OraclePrivKey().Serialize()
+	approverUniqueID := e.reactor.EnclaveInfo().UniqueIDHex()
+
+	// generate transaction message for approval of oracle upgrade
+	encryptedOraclePrivKey, err := encryptOraclePrivKey(oraclePrivKeyBz, oracleUpgrade.NodePubKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encrypt oracle private key: %w", err)
+	}
+
+	approvalMsg := &oracletypes.ApprovalSharingOracleKey{
+		UniqueId:               approverUniqueID,
+		ApproverOracleAddress:  approverAddress,
+		TargetOracleAddress:    targetAddress,
```

Comment:
I think we need to `targetOracleUniqueID` field in `ApprovalSharingOracleKey` struct.
This message should include the uniqueID of the target to be upgraded so that panacea-core can find it.
That is, it is a confirmation that this request is for the current upgrade target.

Reply:
That's good point. It'd be better to include target unique ID and compare it.

Reply:
I'll make an issue in both core and oracle, and then handle it as another PR.
Thank you for your comments.

Reply:
Wait, if the `targetOracleUniqueID` is to enable panacea to find unique ID, there is no need to add it.
panacea can get(find) unique ID of target uniqueID using `GetOracleUpgradeInfo`.
I'll think a little more about whether this is really necessary.

Reply:
In my opinion, as @gyuguen said, it would be better to include `TargetOracleUniqueID` in `ApprovalSharingOracleKey` struct.

It's not a critical point because oracle checks the unique ID from upgrade-oracle event is equal to the current unique ID of upgrade version of oracle.
However, assuming that oracle and core do not believe each other, the unique ID check should also been done in core side.

Reply:
Yes. As you said, the value can be got from OracleUpgradeInfo. 
But I think we will check if the request for Approval is really the uniqueID currently upgrading to.
If it is determined that this validation is not important, the `TargetOracleUniqueID` need not be added. :)

Reply:
I also think it's good to add.
made an [issue](https://github.com/medibloc/panacea-oracle/issues/62) :+1: